### PR TITLE
Terragrunt performance improvements

### LINKED ISSUE #2806: Terragrunt 0.50.15 is 15x slower when your Terragrunt repo has multiple dependencies in my terragrunt.hcl files.
**Describe the bug**
Terragrunt release v0.50.15 is much slower at calculating multiple dependencies.  The commands `terragrunt graph-dependencies`, `terragrunt run-all plan` and `terragrunt run-all apply` are all x15 slower in my scenario.

In releases before v0.50.15, `terragrunt graph-dependencies` takes ~0.5 seconds to execute. With release v0.50.15+ , the same command command takes ~14 seconds to execute.

**To Reproduce**
I put together a example repo to reproduce this issue: 
https://github.com/Simon-TheUser/terragrunt-performance-changes

Run `terragrunt graph-dependencies` from the /iac folder in the project..

Sample hcl:
```hcl
include "root" {
  path   = find_in_parent_folders("root.hcl")
  expose = true
}

dependencies {
  paths = [
    find_in_parent_folders("vm_setup_post-cloud-init"),
    find_in_parent_folders("vm_setup_ssh")
    ]
}

dependency "vm" {
  config_path = find_in_parent_folders("vm_instance/terragrunt.hcl")
}

inputs = {
  ip_address  = dependency.vm.outputs.vm_ip
  vm_uuid     = dependency.vm.outputs.vm_uuid
}
```

**Expected behavior**
A performance losses are sometimes necessary, but I would not expect a x15 performance drop.

**Nice to have**
Here is the time it takes to run 
```bash
$ ./03-print-results.sh
Testing terragrunt graph-dependencies version 0.43.3
real    0m 0.66s
user    0m 0.85s
sys     0m 0.09s

Testing terragrunt graph-dependencies version 0.50.13
real    0m 0.58s
user    0m 0.75s
sys     0m 0.08s

Testing terragrunt graph-dependencies version 0.50.14
real    0m 0.58s
user    0m 0.79s
sys     0m 0.05s

Testing terragrunt graph-dependencies version 0.50.15
real    0m 14.88s
user    0m 20.01s
sys     0m 2.21s

Testing terragrunt graph-dependencies version 0.50.16
real    0m 14.58s
user    0m 19.89s
sys     0m 1.99s

Testing terragrunt graph-dependencies version 0.51.7
real    0m 14.23s
user    0m 19.17s
sys     0m 2.19s

Testing terragrunt graph-dependencies version 0.53.6
real    0m 11.46s
user    0m 11.23s
sys     0m 1.20s
```

**Versions**
- Terragrunt version: 0.50.15
- Terraform version: 1.3.10 *not important.
- Environment details (Ubuntu 20.04, Windows 10, etc.): Running inside a Podman container. Base image is alpine:3.18.3

### LINKED ISSUE #2980: Extremely poor performance in v0.55.5+
**Describe the bug**
It seems that the performance since v0.55.6 is very bad. A simple `terragrunt apply` without caches (that is with `.terragrunt-cache` and `.terraform.lock.hcl` recursively removed) is taking ages.  
What took about a minute on v0.55.5 didn't even finish after 10 minutes on v0.55.6.

I then downgraded back to v0.55.5 and it worked again.
It seems like it is getting caught in some kind of cyclic redundancy as it keeps looping over the same modules over and over again ("Evaluated X locals (remaining XX)")

**Versions**
- Terragrunt version: v0.55.6
- Terraform version: 1.7.4
- Environment details Ubuntu 23.10

**Additional context**
- Cause could be the changes done in #2957 to fix #2955
- Other users are reporting performance issues as well in #2806

### PR DESCRIPTION
**Pull request by denis256**
## Description

Included changes:
* added more telemetry steps collection
* replaced typed cache implementation with generic version
* added cache for parsed HCL files
* added cache for parsed terraform modules

Before this changes:
image
```
real    0m20.526s
user    0m25.007s
sys     0m1.258s
```

After changes from this PR:
image

```
real    0m2.417s
user    0m2.206s
sys     0m0.268s
```

Fixes #2980.
Fixes #2806.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

### CODE REVIEW

**Comment by Simon-TheUser** [comment]
Nice work!

Testing with Terragrunt version terragrunt.v0.55.15
real	0m 24.94s
user	0m 30.44s
sys	0m 2.04s

Testing with Terragrunt version terragrunt.v0.55.15-31-gcf820b0ed84f
real	0m 5.35s
user	0m 3.27s
sys	0m 0.26s

(I stored the compete output of my test here: https://github.com/Simon-TheUser/terragrunt-performance-changes/blob/main/notebook.ipynb)

File: configstack/module.go [hunk 1]
```diff
@@ -57,26 +59,85 @@ func ResolveTerraformModules(terragruntConfigPaths []string, terragruntOptions *
...
+		crossLinkedModules = result
+		return nil
+	})
 	if err != nil {
 		return nil, err
 	}
 
-	includedModules := flagIncludedDirs(crossLinkedModules, terragruntOptions)
+	var includedModules []*TerraformModule
+	err = telemetry.Telemetry(terragruntOptions, "flag_included_dirs", map[string]interface{}{
```
Comment (brikis98) [inline configstack/module.go:109]: Good to see lots of extra telemetry 👍 

Was it useful in tracking down these perf issues?

**Review by brikis98 — APPROVED** [review]
LGTM! Very nice perf improvement 👍

File: configstack/module.go (hunk 1 above)
Comment (denis256) [reply configstack/module.go:109]: Yes, it was helpful to find which part consumes more time during the execution

**Comment by julienbonastre** [comment]
Still occurring ...

**Pull request merged**